Add a blog post with 2024 day 5 solution (#610)
scalacenter/scala-advent-of-code: docs/2024/puzzles/day05.md

# Day 5: Print Queue

by [@KacperFKorban](https://github.com/KacperFKorban)

## Puzzle description

https://adventofcode.com/2024/day/5

> ## Solution Sumary

Reviewer: ```suggestion
## Solution Summary
```

We can treat the data as a graph, where:
- the ordering rules represent directed edges in the graph
- each update represents a subset of the nodes

As a common part of the solution, we will:
- parse the input into a list of ordering rules `(Int, Int)` and a list of updates `List[List[Int]]`
- represent the rules as an adjacency list `Map[Int, List[Int]]`

1. **Part 1**:

>    - For every update, we iterate over its elements (while keeping track of the visited nodes) and for each node check that **none** of its neighbors were visited before it.

Reviewer: ```suggestion
   - For every update, we iterate over its elements (while keeping track of the visited nodes) and for each node check that **none** of its successors were visited before it.
```

  def rec(update: List[Int], visited: Set[Int] = Set.empty): Boolean = update match
    case Nil => true
    case updateNo :: rest =>
      !rules.getOrElse(updateNo, List.empty).exists(visited.contains)
        && rec(rest, visited + updateNo)
  rec(update)
```

another alternative is using `boundary`-`break` with a `for`:

```scala
def isValid(rules: Map[Int, List[Int]])(update: List[Int]): Boolean =
  boundary:
    var visited = Set.empty[Int]
    for updateNo <- update do
      visited += updateNo
      if rules.getOrElse(updateNo, List.empty).exists(visited.contains) then
        break(false)
    true
```

or a `forall` with a local mutable state:

```scala
def isValid(rules: Map[Int, List[Int]])(update: List[Int]): Boolean =
  var visited = Set.empty[Int]
  update.forall { updateNo =>
    visited += updateNo
    !rules.getOrElse(updateNo, List.empty).exists(visited.contains)
  }
```

Using the `isValid` function, we can filter the updates and compute the middle number of the valid updates:

```scala
def part1(input: String) =
  val (rules, updates) = parseRulesAndupdates(input)
  updates.filter(isValid(rules)).map(us => us(us.size / 2)).sum
```

## Part 2

> We start Part 2 by parsing and filtering the updates that violate the ordering rules, very similarily to Part 1:

Reviewer: ```suggestion
We start Part 2 by parsing and filtering the updates that violate the ordering rules, very similarly to Part 1:
```

```scala
val (rules, updates) = parseRulesAndupdates(input)
val invalidupdates = updates.filter(!isValid(rules)(_))
```

Next, to fix a single update, we first construct local adjacency lists for the relevant nodes and an inverse adjacency list to keep track of the incoming edges:

```scala
def fixUpdate(update: List[Int]): List[Int] =
  val relevantRules = rules
    .filter((k, vs) => update.contains(k) && vs.exists(update.contains))
    .mapValues(_.filter(update.contains)).toMap
  val prevsMap = relevantRules
    .map { case (k, vs) => vs.map(_ -> k) }
    .flatten.groupMap(_._1)(_._2)
```

> The `relevantRules` are only those that only use the nodes from `update`, and the `prevsMap` is a map from a node to its incoming edges.

Reviewer: ```suggestion
The `relevantRules` are only those that only use the nodes from `update`, and the `prevsMap` is a map from a node to its direct predecessors.
```
